### PR TITLE
Added patch envelope support

### PR DESCRIPTION
**Pull request by shrikanthb-zededa**
Added support to manage patch envelope through Zededa terraform provider.
Auto generated client, models, schemas files for patch envelope from the latest app_instance swagger.json file. Added resources file for patch envelope operations. Made necessary changes to auto generated files for it to work with resources file.
It supports creating/updating patch envelopes and attaching/detaching it from app instance.
Test case added for patch envelope management through terraform-provider. 
Documentation updated to include the patch envelope resources and data-sources files.
main.tf formatted with `terraform fmt`.

### CODE REVIEW

File: resources/patch_envelope.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+		}
+	}
+
+	// note, we need to set the ID in any case, even GetByName endpoint seems to require the ID
+	// but doesn't return any error if it's not set.
+	d.SetId(responseData.ObjectID)
+
+	// the zedcloud API does not return the partially updated object but a custom response.
+	// thus, we need to fetch the object and populate the state.
+	if errs := readPatchEnvelopeByName(ctx, d, m); err != nil {
```
Comment (xyuria-zededa) [inline resources/patch_envelope.go]: It's better to rewrite this code as the following:
`if diags := readPatchEnvelopeByName(ctx, d, m); diags.HasError() {
		return diags
	}`
The reason is that the variable err doesn't belong to this scope and  doesn't check correctness of the readPathEnvelopeByName response.

File: resources/patch_envelope.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+		}
+	}
+
+	// note, we need to set the ID in any case, even GetByName endpoint seems to require items
+	// but doesn't return any error if it's not set.
+	d.SetId(responseData.ObjectID)
+
+	// the zedcloud API does not return the partially updated object but a custom response.
+	// thus, we need to fetch the object and populate the state.
+	if errs := readPatchEnvelopeByName(ctx, d, m); err != nil {
```
Comment (xyuria-zededa) [inline resources/patch_envelope.go]: The code better be rewritten as mentioned in the comment for CreatePathEnvelope function.

File: resources/patch_reference_update.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		}
+	}
+
+	// note, we need to set the ID in any case, even GetByName endpoint seems to require the ID
+	// but doesn't return any error if it's not set.
+	d.SetId(responseData.ObjectID)
+
+	// the zedcloud API does not return the partially updated object but a custom response.
+	// thus, we need to fetch the object and populate the state.
+	if errs := readPatchEnvelopeByName(ctx, d, m); err != nil {
```
Comment (xyuria-zededa) [inline resources/patch_reference_update.go]: It's better to rewrite this code as the following:
`if diags := readPatchEnvelopeByName(ctx, d, m); diags.HasError() {
		return diags
	}`
The reason is that the variable err doesn't belong to this scope and  doesn't check correctness of the readPathEnvelopeByName response.

File: resources/utils/helper_functions.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package utils
+
+// remove an item from array
+func Remove(s []string, r string) []string {
```
Comment (xyuria-zededa) [inline resources/utils/helper_functions.go]: Do we use this helper function somewhere? If not, please remove it.

File: main.tf [hunk 5]
```diff
@@ -1911,43 +1922,104 @@ resource "zedcloud_application_instance"  "test_tf_provider" {
...
-		# restart {
-	    	# counter = 2
-		# }
-		tags = {
-        "tag-key-1" = "tag-value-1"
+  }
+  artifacts {
+    format = "OpaqueObjectCategoryExternalBinary"
+    binary_artifact {
+      image_id = resource.zedcloud_image.test_tf_provider.id
```
Comment (xyuria-zededa) [inline main.tf]: Here and below we don't need to use 'resource' when specifying a dependency. Just `zedcloud_image.test_tf_provider.id`. Please correct it when you have time.

**Review by xyuria-zededa — CHANGES_REQUESTED** [review]
Added a couple of comments. Please take a look when you have time.

File: resources/patch_envelope.go (hunk 1 above)
Comment (shrikanthb-zededa) [reply resources/patch_envelope.go]: updated.

File: resources/patch_envelope.go (hunk 2 above)
Comment (shrikanthb-zededa) [reply resources/patch_envelope.go]: updated.

File: resources/patch_reference_update.go (hunk 3 above)
Comment (shrikanthb-zededa) [reply resources/patch_reference_update.go]: removed reading patch envelope after updating patch envelope reference as it is not useful.

File: resources/utils/helper_functions.go (hunk 4 above)
Comment (shrikanthb-zededa) [reply resources/utils/helper_functions.go]: removed it

File: main.tf (hunk 5 above)
Comment (shrikanthb-zededa) [reply main.tf]: corrected it

**Review by xyuria-zededa — APPROVED** [review]
LGTM

**Pull request merged**